include/maliput/common/profiler.h

```diff
@@ -0,0 +1,71 @@
+// BSD 3-Clause License
+//
+// Copyright (c) 2023, Woven Planet.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+/// @file Wraps ignition common's profiler behind maliput macros.
+///       This allows us to enable/disable profiling at compile time and avoid adding a dependency on ignition common.
+
+#ifndef MALIPUT_PROFILER_ENABLE
+/// Always set this variable to some value
+#define MALIPUT_PROFILER_ENABLE 0
+#endif
+
+#if MALIPUT_PROFILER_ENABLE
+
+#define IGN_PROFILER_ENABLE 1
+#include <ignition/common/Profiler.hh>
+
+/// \brief Set name of profiled thread
+#define MALIPUT_PROFILE_THREAD_NAME(name) IGN_PROFILE_THREAD_NAME(name)
+/// \brief Log profiling text, if supported by implementation
+#define MALIPUT_PROFILE_LOG_TEXT(name) IGN_PROFILE_LOG_TEXT(name)
+/// \brief Being profiling sample
+#define MALIPUT_PROFILE_BEGIN(name) IGN_PROFILE_BEGIN(name)
+/// \brief End profiling sample
+#define MALIPUT_PROFILE_END() IGN_PROFILE_END()
+
+/// \brief Convenience wrapper for scoped profiling sample. Use MALIPUT_PROFILE
+#define MALIPUT_PROFILE_L(name, line) IGN_PROFILE_L(name, line)
+/// \brief Scoped profiling sample. Sample will stop at end of scope.
+#define MALIPUT_PROFILE(name) IGN_PROFILE(name)
```

Comment:
Is it possible to use `__FUNC__` / `__PRETTY_FUNCTION__` (even better but not sure if it is available in ) ?

That will allow to reduce the cost to implement the profiling everywhere, meaning that we don't need to review spelling mistakes and so on and the compiler will do it for us.

Reply:
Maybe we could create a new macro like 
```cpp
#define MALIPUT_PROFILE_FUNC() IGN_PROFILE(__FUNCTION__)
```
- Using __FUNCTION__ only gives you the name of the method, without carrying about namespacing. So I wouldn't use it
- Using __PRETTY_FUNCTION__ is possible, I tried it using compiler explorer for both gcc and clang. The firm of the method is quite complete: namespaces + return type + argument. The only issue here is that when using the profiler UI in order to analyze data it isn't that user friendly with long names, it is basically a mess. (I can try this out and post a picture here using the profiler)


Finally, we are using the profiler for analyzing entire method execution, however, it could be used por different part of a method. For example, check this usecase in [gz-sim::simulationrunner](https://github.com/gazebosim/gz-sim/blob/9a36bc3c1620b047ef45571afc817cfdba1e41fb/src/SimulationRunner.cc#L548-L588)

Reply:
I like the macro `#define MALIPUT_PROFILE_FUNC() IGN_PROFILE(__FUNCTION__)`

Reply:
I think it is OK, but I would use it everywhere we are using the name of the function only to avoid spelling mistakes. Also, the macro could be (I hope the compiler shakes my right hand):

```C++
#define MALIPUT_PROFILE_FUNC()  IGN_PROFILE(__FUNCTION__)
#define MALIPUT_PROFILE_FUNC_MSG(msg) IGN_PROFILE(__FUNCTION_": "#msg)
```

Reply:
Out of curiosity, are these `__FUNCTION__` and `__PRETTY_FUNCTION` in any of the standards? Never saw them before. If they are not, not sure guys how you deal with this platform constraints decisions in maliput.

Reply:
> Out of curiosity, are these `__FUNCTION__` and `__PRETTY_FUNCTION` in any of the standards? Never saw them before. If they are not, not sure guys how you deal with this platform constraints decisions in maliput.

Good question, `__FUNCTION__` is part of the C++ standard already. Regarding `__PRETTY_FUNCTION__` apparently it is not in the standard (at least yet) however gcc and clang have support for this. Given that we state to support both gcc and clang compilers, I think that there is no harm in using them.

Reply:
> ```c++
> #define MALIPUT_PROFILE_FUNC_MSG(msg) IGN_PROFILE(__FUNCTION_": "#msg)
> ```

The issue with `#define MALIPUT_PROFILE_FUNC_MSG(msg) IGN_PROFILE(__FUNCTION_": "#msg)` is the argument:
IGN_PROFILER is expecting a `char *` and :
 - `__FUNCTION__": "#msg`
 - `__FUNCTION__ + ": " + #msg`
 aren't correct.

 There is an alternative of concatenating as a string and converting to const char * but making this conversion/operation every time that the method is called isn't ideal.
 `#define MALIPUT_PROFILE_FUNC_MSG(msg) IGN_PROFILE((static_cast<std::string>(__FUNCTION__) + std::string(": ") + static_cast<std::string>(msg)).c_str())
`


--------------------------------------------


I replaced all the occurrences with `MALIPUT_PROFILE_FUNC()` Check [209cc9d](https://github.com/maliput/maliput/pull/538/commits/209cc9d849d63e2333a5914c486902f53a7c7670)
![image](https://user-images.githubusercontent.com/53065142/214701722-fbe36346-095c-4a16-bdce-29b3154c8b66.png)

Reply:
Good catch. Let's keep it as is then.